### PR TITLE
Endpoint startup experience

### PR DESCRIPTION
**Pull request by JamesNK**
https://github.com/aspnet/Routing/issues/909

- [x] Merge MVC changes to get MapApplication in non-feature branch (cause of build failure)
- [x] Update F# templates

### CODE REVIEW

File: src/Microsoft.DotNet.Web.ProjectTemplates/content/RazorPagesWeb-CSharp/Startup.cs [hunk 1]
```diff
@@ -156,13 +156,18 @@ public void Configure(IApplicationBuilder app, IHostingEnvironment env)
...
+            });
+
             app.UseCookiePolicy();
 
 #if (OrganizationalAuth || IndividualAuth)
             app.UseAuthentication();
-
 #endif
-            app.UseMvc();
+            app.UseAuthorization();
```
Comment (rynowak) [inline src/Microsoft.DotNet.Web.ProjectTemplates/content/RazorPagesWeb-CSharp/Startup.cs:170]: Should this be in the block with `UseAuthentication()`? What would this do if you don't have authn?

Comment (JamesNK) [reply src/Microsoft.DotNet.Web.ProjectTemplates/content/RazorPagesWeb-CSharp/Startup.cs:170]: Imagine you have an MVC app without AuthN. You add `[Authorize]` to an action. Visiting that action returns an error from MVC that AuthN is not setup.

Now add a non-MVC endpoint, e.g. a `MapGet`. You add `RequiresAuthorization` to it. Should it throw an error saying that AuthN is not setup or should it allow requests to access the endpoint with AuthZ metadata silently?

Crazy idea: If we don't like always having UseAuthorization, could authz be baked into Http.Abstractions? i.e. it is always checked before executing the endpoint.

File: src/Microsoft.DotNet.Web.ProjectTemplates/content/RazorPagesWeb-CSharp/Startup.cs [hunk 2]
```diff
@@ -156,13 +156,18 @@ public void Configure(IApplicationBuilder app, IHostingEnvironment env)
 
 #endif
             app.UseStaticFiles();
+
+            app.UseEndpointRouting(routes =>
+            {
+                routes.MapApplication();
```
Comment (JamesNK) [inline src/Microsoft.DotNet.Web.ProjectTemplates/content/RazorPagesWeb-CSharp/Startup.cs:162]: Change to MapAssembly for web apis
* Controllers - controller type
* Razor pages - application parts - follow the assembly for embedded pages

**Comment by davidfowl** [comment]
Aren't we renaming this thing to UseRouting?

File: src/Microsoft.DotNet.Web.ProjectTemplates/content/EmptyWeb-FSharp/Startup.fs [hunk 3]
```diff
@@ -18,4 +18,6 @@ type Startup() =
         if env.IsDevelopment() then 
             app.UseDeveloperExceptionPage() |> ignore
 
-        app.Run(fun context -> context.Response.WriteAsync("Hello World!")) |> ignore
\ No newline at end of file
+        app.UseRouting(fun routing ->
```
Comment (rynowak) [inline src/Microsoft.DotNet.Web.ProjectTemplates/content/EmptyWeb-FSharp/Startup.fs:21]: I agree. Routing **is** fun

File: src/Microsoft.DotNet.Web.ProjectTemplates/content/RazorPagesWeb-CSharp/Startup.cs (hunk 2 above)
Comment (rynowak) [reply src/Microsoft.DotNet.Web.ProjectTemplates/content/RazorPagesWeb-CSharp/Startup.cs:162]: I thought we wanted MapAssembly for everything? Are there bugs blocking this?

**Comment by JamesNK** [comment]
> Aren't we renaming this thing to UseRouting?

Done.

Hmmm, updating dependencies causes the PR build to fail when running integration tests:

```
   Running ASP.NET application...
   ==> C:\Users\James\.dotnet\x64\dotnet.exe exec bin/Debug/netcoreapp3.0/AspNet.Template.5240ad7043d54e06a1420ea407dd7e06.dll [C:\Development\Source\Templating\test\Templates.Test\bin\Debug\netcoreapp3.0\TestTemplates\AspNet.Template.5240ad7043d54e06a1420ea407dd7e06]
   Waiting until ASP.NET application is accepting connections...
   [ERROR]
   [ERROR] Unhandled Exception: System.TypeInitializationException: The type initializer for 'System.AppContext' threw an exception. ---> System.IO.FileNotFoundException: Could not load file or assembly 'Microsoft.AspNetCore.Mvc.Core, Version=3.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60'. The system cannot find the file specified.
   [ERROR]    at System.ModuleHandle.ResolveType(RuntimeModule module, Int32 typeToken, IntPtr* typeInstArgs, Int32 typeInstCount, IntPtr* methodInstArgs, Int32 methodInstCount, ObjectHandleOnStack type)
   [ERROR]    at System.ModuleHandle.ResolveTypeHandleInternal(RuntimeModule module, Int32 typeToken, RuntimeTypeHandle[] typeInstantiationContext, RuntimeTypeHandle[] methodInstantiationContext)
   [ERROR]    at System.Reflection.RuntimeModule.ResolveType(Int32 metadataToken, Type[] genericTypeArguments, Type[] genericMethodArguments)
   [ERROR]    at System.Reflection.CustomAttribute.FilterCustomAttributeRecord(CustomAttributeRecord caRecord, MetadataImport scope, RuntimeModule decoratedModule, MetadataToken decoratedToken, RuntimeType attributeFilterType, Boolean mustBeInheritable, ListBuilder`1& derivedAttributes, RuntimeType& attributeType, IRuntimeMethodInfo& ctor, Boolean& ctorHasParameters, Boolean& isVarArg)
   [ERROR]    at System.Reflection.CustomAttribute.AddCustomAttributes(ListBuilder`1& attributes, RuntimeModule decoratedModule, Int32 decoratedMetadataToken, RuntimeType attributeFilterType, Boolean mustBeInheritable, ListBuilder`1& derivedAttributes)
   [ERROR]    at System.Reflection.CustomAttribute.GetCustomAttributes(RuntimeModule decoratedModule, Int32 decoratedMetadataToken, Int32 pcaCount, RuntimeType attributeFilterType)
   [ERROR]    at System.Attribute.GetCustomAttributes(Assembly element, Type attributeType, Boolean inherit)
   [ERROR]    at System.Attribute.GetCustomAttribute(Assembly element, Type attributeType, Boolean inherit)
   [ERROR]    at System.Reflection.CustomAttributeExtensions.GetCustomAttribute[T](Assembly element)
   [ERROR]    at System.AppDomainSetup.get_TargetFrameworkName()
   [ERROR]    at System.AppContextDefaultValues.ParseTargetFrameworkName(String& identifier, String& profile, Int32& version)
   [ERROR]    at System.AppContextDefaultValues.PopulateDefaultValues()
   [ERROR]    --- End of inner exception stack trace ---
   [ERROR]    at System.StartupHookProvider.ProcessStartupHooks()
```

**Comment by JamesNK** [comment]
Moved to aspnetcore

**Pull request closed without merging**